### PR TITLE
webui: support cancelation

### PR DESCRIPTION
**Pull request by bakkot**
This is on top of https://github.com/lstein/stable-diffusion/pull/181, which is itself on top of https://github.com/lstein/stable-diffusion/pull/180. (Github does not have a good way of staging multiple PRs.) Changes in this PR are just the "support cancelation" commit. This PR works even without https://github.com/lstein/stable-diffusion/pull/181, but it's a lot more useful with it.

This adds a button next to the progress bar which allows you to cancel the in-progress generation - e.g. if you asked for 5 images and then discovered after the first 2 that you want to tweak the prompt, or if the partial rendering from https://github.com/lstein/stable-diffusion/pull/180 was enough to know it wasn't what you were looking for. It accomplishes this by throwing an exception in the per-step callback and catching it outside the call to `prompt2image`, which, while a bit gross, is fundamentally the same thing as hitting ctrl-c and catching the KeyboardInterrupt. That's [what the REPL does](https://github.com/lstein/stable-diffusion/blob/9a8cd9684e0236b22befd14fed26123f75f20f8d/scripts/dream.py#L236), so it should be ok.

Cancelation does need to wait for the current step to finish rather than actually being instant, but it's only a second or two of delay, and I didn't really want to get into dealing with actual interrupts or [deeper magic](https://stackoverflow.com/a/61528202).

<details>
<summary>screenshot of button</summary>

Screen Shot 2022-08-29 at 12 23 12 AM
</details>

### CODE REVIEW

**Comment by bakkot** [comment]
I'll rebase this after https://github.com/lstein/stable-diffusion/pull/186 and https://github.com/lstein/stable-diffusion/pull/181 are merged.

**Review by lstein — APPROVED** [review]
Good addition to the previus PR. Hopefully we can get both in soon. There are source conflicts, but they don't look too daunting.

**Comment by blessedcoolant** [comment]
I tested this PR. It's working fine. I can approve it if you want @lstein ... Both the streaming and the cancelation are working as expected.

**Comment by lstein** [comment]
> I tested this PR. It's working fine. I can approve it if you want @lstein ... Both the streaming and the cancelation are working as expected.
@blessedcoolant I'm trying to be disciplined about the code reviews, so if you could do a formal approval that would be great. Meanwhile I'll clean up the conflicts.

**Comment by lstein** [comment]
@bakkot , we've run into the CRLF issue again. Could you fix the line endings and rebase?

**Comment by bakkot** [comment]
Rebased.

File: ldm/dream/server.py [hunk 1]
```diff
@@ -133,41 +148,46 @@ def image_progress(sample, step):
...
-                                    sampler_name    = sampler_name,
-                                    gfpgan_strength=gfpgan_strength,
-                                    upscale         = upscale,
-                                    step_callback=image_progress,
-                                    image_callback=image_done)
-
-            # Remove the temp file
-            os.remove("./img2img-tmp.png")
+        try:
+            if initimg is None:
```
Comment (bakkot) [inline ldm/dream/server.py:152]: This change is mostly whitespace (I added a couple `try`s), so I recommend viewing the diff [with whitespace suppressed](https://stackoverflow.com/questions/37007300/how-to-ignore-whitespace-in-github-when-comparing/51755490#51755490).

**Comment by lstein** [comment]
@blessedcoolant indicated review approval in comments, so going ahead with merge.

**Pull request merged**